rstar/src/primitives/geomwithdata.rs

```diff
@@ -0,0 +1,175 @@
+///
+/// # Examples
+/// ```
+/// use rstar::{RTree, PointDistance};
+/// use rstar::primitives::GeomWithData;
+///
+/// type RestaurantLocation = GeomWithData<&'static str, [f64; 2]>;
+///
+/// let mut restaurants = RTree::new();
+/// restaurants.insert(RestaurantLocation::new("Pete's Pizza Place", [0.3, 0.2]));
+/// restaurants.insert(RestaurantLocation::new("The Great Steak", [-0.8, 0.0]));
+/// restaurants.insert(RestaurantLocation::new("Fishy Fortune", [0.2, -0.2]));
+///
+/// let my_location = [0.0, 0.0];
+///
+/// // Now find the closest restaurant!
+/// let place = restaurants.nearest_neighbor(&my_location).unwrap();
+/// println!("Let's go to {}", place.get_data());
+/// println!("It's really close, only {} miles", place.distance_2(&my_location));
+/// ```
+/// ```
+/// // We can use any primitive type
+/// use rstar::{RTree, PointDistance};
+/// use rstar::primitives::GeomWithData;
+/// use rstar::primitives::Line;
+///
+/// let mut lines = RTree::new();
+/// lines.insert(GeomWithData::new(
+///     "Pete's Pizza Place",
+///     Line::new([0.0, 0.0], [1.0, 1.0]),
+/// ));
+/// lines.insert(GeomWithData::new(
+///     "The Great Steak",
+///     Line::new([0.0, 0.0], [-1.0, 1.0]),
+/// ));
+///
+/// let my_location = [0.0, 0.0];
+///
+/// // Now find the closest line
+/// let place = lines.nearest_neighbor(&my_location).unwrap();
+/// println!("Let's go to {}", place.get_data());
+/// println!(
+///     "It's really close, only {} miles",
+///     place.get_geom().distance_2(&my_location)
+/// );
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub struct GeomWithData<T, R>(T, R);
+
+impl<T, R: RTreeObject> RTreeObject for GeomWithData<T, R> {
+    type Envelope = R::Envelope;
+
+    fn envelope(&self) -> Self::Envelope {
+        self.1.envelope()
+    }
+}
+
+impl<T, R: PointDistance> PointDistance for GeomWithData<T, R> {
+    fn distance_2(
+        &self,
+        point: &<Self::Envelope as Envelope>::Point,
+    ) -> <<Self::Envelope as Envelope>::Point as Point>::Scalar {
+        self.1.distance_2(point)
+    }
+
+    fn contains_point(&self, p: &<Self::Envelope as Envelope>::Point) -> bool {
+        self.1.contains_point(p)
+    }
+
+    fn distance_2_if_less_or_equal(
+        &self,
+        point: &<Self::Envelope as Envelope>::Point,
+        max_distance_2: <<Self::Envelope as Envelope>::Point as Point>::Scalar,
+    ) -> Option<<<Self::Envelope as Envelope>::Point as Point>::Scalar> {
+        self.1.distance_2_if_less_or_equal(point, max_distance_2)
+    }
+}
+
+impl<T, P> GeomWithData<T, P> {
+    /// Creates a new [GeomWithData] with the provided data.
+    pub fn new(data: T, geom: P) -> Self {
+        GeomWithData(data, geom)
+    }
+
+    /// Get a reference to the container's data
+    pub fn get_data(&self) -> &T {
+        &self.0
+    }
+
+    /// Set the container's data
+    pub fn set_data(&mut self, new_value: T) {
```

Comment:
Given that we are trying to provide `get`, `set` and `replace`, we might as well make `data` a `pub` field?  Something like:

```rust
pub struct GeomWithData<T, R>{
  geom: T,
  pub data: R,
}
```

Feels a bit more ergonomic to access the data.

Reply:
(I concur with the above, but even if we don't, I think it would be more idiomatic to provide something like `fn data_mut(&mut self) -> &mut T` instead of by value setter and replace method.)

Reply:
My only reservation about making `data` public is that it's then public forever. In practice, it's unlikely that we'll be altering this struct in such a way as to restrict or otherwise redirect `data`, but a setter is more conservative. In any case, it's pub for now.

Reply:
I'll add one more voice...

```
fn data(&self) -> &T;
fn data_mut(&self) -> &mut T;
```

from: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter
> With a few exceptions, the get_ prefix is not used for getters in Rust code.

Once you have a &mut there's not really a reason to have a set *and* a replace, plus then you can also do fun things like `swap` and `take` if you want them.

Reply:
oh sorry, I think I erroneously commented on an old diff here. I'm also ok with the `pub data: T,` solution you've adopted. 👍